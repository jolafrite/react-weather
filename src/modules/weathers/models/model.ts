export interface IFetchWeather extends IWeather {}

export interface IWeather {
  city: {
    id: number;
    name: string;
    coord: {
      lon: number;
      lat: number;
    };
    country: string;
  };
  cnt: number;
  list: IWeatherList[];
};

export interface IWeatherList {
  dt: number;
  main: {
    temp: number;
    temp_min: number;
    temp_max: number;
    pressure: number;
    humidity: number;
    sea_level: number;
    grnd_level: number;
    temp_kf: number;
  };
  weather: IWeatherListWeather[];
  clouds: {
    all: number;
  };
  wind: {
    speed: number;
    deg: number;
  };
  dt_txt: string;
};

export interface IWeatherListWeather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

export interface IWeatherLast5DaysForecast {
  [date: string]: IWeatherDailyForecast;
}

export interface IWeatherDailyForecast {
  temp: number;
  temp_max: number;
  temp_min: number;
  hourly: {
    [time: string]: IWeatherList;
  };
}
