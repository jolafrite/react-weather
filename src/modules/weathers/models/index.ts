export * from './weather.fixtures';
export * from './model';
