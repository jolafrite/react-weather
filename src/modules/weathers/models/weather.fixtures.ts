import { useFixtureFactory } from "../../../common/utils/test";
import { IWeather } from "./model";
import { ISearchWeather } from "../../../fetch/search-weather/model";

export const weatherFixtures = useFixtureFactory<IWeather>({
  city: {
    id: 1851632,
    name: "Shuzenji",
    coord: { lon: 138.933334, lat: 34.966671 },
    country: "JP"
  },
  cnt: 38,
  list: [
    {
      dt: 1406106000,
      main: {
        temp: 298.77,
        temp_min: 298.77,
        temp_max: 298.774,
        pressure: 1005.93,
        sea_level: 1018.18,
        grnd_level: 1005.93,
        humidity: 87,
        temp_kf: 0.26
      },
      weather: [
        { id: 804, main: "Clouds", description: "overcast clouds", icon: "04d" }
      ],
      clouds: { all: 88 },
      wind: { speed: 5.71, deg: 229.501 },
      dt_txt: "2014-07-23 09:00:00"
    }
  ]
});

export const searchWeatherFixtures = useFixtureFactory<ISearchWeather>({
  id: 2643743,
  name: "London",
  coord: { lat: 51.5085, lon: -0.1258 },
  main: {
    temp: 284.31,
    pressure: 1010,
    humidity: 87,
    temp_min: 283.15,
    temp_max: 285.93,
    sea_level: 1,
    grnd_level: 2,
    temp_kf: 3,
  },
  dt: 1560233226,
  wind: {
    speed: 1.5,
    deg: 150
  },
  sys: {
    country: "GB"
  },
  rain: {
    "1h": 3,
    "3h": 4,
  },
  snow: {
    "1h": 1,
    "3h": 2,
  },
  clouds: { all: 90 },
  weather: [
    { id: 500, main: "Rain", description: "light rain", icon: "10d" },
    { id: 701, main: "Mist", description: "mist", icon: "50d" }
  ]
});
