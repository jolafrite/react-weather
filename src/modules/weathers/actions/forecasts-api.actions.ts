import { ActionType } from "../../../common/utils/redux";
import { IWeatherForecast } from "../models";

export const LOAD_ONE_SUCCESS = '[Forecast API] Load Forecast Success';

export const loadOneSuccess = (forecast: IWeatherForecast) => {
  return  {
    type: LOAD_ONE_SUCCESS,
    forecast
  } as const;
};

const all = {
  loadOneSuccess,
};

export type ForecastApiActionsUnion = ActionType<typeof all>;
