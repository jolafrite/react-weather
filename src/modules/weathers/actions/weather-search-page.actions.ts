import { IWeather } from "../models";
import { ActionType } from "../../../common/utils/redux";

export const RESET = '[Weather Search Page] Reset Search Weather';
export const SEARCH = '[Weather Search Page] Search Weather';
export const SEARCH_SUCCESS = '[Weather Search Page] Search Weathers Success';
export const SEARCH_ERROR = '[Weather Search Page] Search Weathers Error';

export const reset = () => {
  return {
    type: RESET,
  } as const;
};

export const search = (query: string) => {
  return {
    type: SEARCH,
    query,
  } as const;
};

export const searchSuccess = (weathers: IWeather[]) => {
  return {
    type: SEARCH_SUCCESS,
    weathers,
  } as const;
};

export const searchError = (error: Error) => {
  return {
    type: SEARCH_ERROR,
    error
  } as const;
};

const all = {
  reset,
  search,
  searchSuccess,
  searchError,
};

export type WeatherSearchPageActionsUnion = ActionType<typeof all>;
