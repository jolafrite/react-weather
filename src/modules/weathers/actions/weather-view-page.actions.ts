import { ActionType } from "../../../common/utils/redux";
import { IWeather, IWeatherForecast } from "../models";

export const SELECT_ONE = '[Weather View Page] Select Weather';
export const LOAD_ONE = '[Weather View Page] Load Weather';
export const LOAD_ONE_SUCCESS = '[Weather View Page] Load Weather Success';
export const LOAD_ONE_ERROR = '[Weather View Page] Load Weather Error';

export const selectOne = (id: number) => {
  return {
    type: SELECT_ONE,
    id,
  } as const;
};

export const loadOne = () => {
  return {
    type: LOAD_ONE,
  } as const;
};

export const loadOneSuccess = (weather: IWeather, forecasts?: IWeatherForecast[]) => {
  return {
    type: LOAD_ONE_SUCCESS,
    weather,
    forecasts,
  } as const;
};

export const loadOneError = (error: Error) => {
  return {
    type: LOAD_ONE_ERROR,
    error
  } as const;
};

const all = {
  selectOne,
  loadOne,
  loadOneSuccess,
  loadOneError,
};

export type WeatherViewPageActionsUnion = ActionType<typeof all>;
