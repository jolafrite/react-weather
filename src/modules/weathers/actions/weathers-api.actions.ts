import { IWeather } from "../models";
import { ActionType } from "../../../common/utils/redux";

export const LOAD_ONE_SUCCESS = '[Weather API] Load Weather Success';
export const LOAD_LIST_SUCCESS = '[Weather API] Load Weathers Success';

export const loadOneSuccess = (weather: IWeather) => {
  return {
    type: LOAD_ONE_SUCCESS,
    weather,
  } as const;
};

export const loadListSuccess = (weathers: IWeather[]) => {
  return  {
    type: LOAD_LIST_SUCCESS,
    weathers
  } as const;
};

const all = {
  loadOneSuccess,
  loadListSuccess,
};

export type WeatherApiActionsUnion = ActionType<typeof all>;
