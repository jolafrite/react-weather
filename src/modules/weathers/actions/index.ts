import * as WeatherSearchPageActions from './weather-search-page.actions';
import * as WeatherViewPageActions from './weather-view-page.actions';
import * as WeatherApiActions from './weathers-api.actions';

export {
  WeatherSearchPageActions,
  WeatherViewPageActions,
  WeatherApiActions,
};
