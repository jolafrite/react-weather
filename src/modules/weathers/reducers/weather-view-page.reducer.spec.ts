import { reducer, getError, getLoading, getId } from "./weather-view-page.reducer";
import { WeatherViewPageActions } from "../actions";
import { weatherFixtures } from "../models";
import { getLoaded, initialState } from "./weather-view-page.reducer";

describe('reducer: Weather View Page', () => {

  describe('action:', () => {

    describe('undefined', () => {
      test('should result the default state', () => {
        expect(
          reducer(undefined, {} as any)
        ).toMatchSnapshot();
      });
    });

    describe('ViewPageActions.selectOne', () => {
      test('should reset set state', () => {
        const state = {
          ...initialState,
          status: 'success',
          error: new Error('an error')
        };

        expect(
          reducer(state, WeatherViewPageActions.selectOne(1))
        ).toMatchSnapshot();
      });

      test('should set the id', () => {
        expect(
          reducer(undefined, WeatherViewPageActions.selectOne(1))
        ).toMatchSnapshot();
      });
    });

    describe('ViewPageActions.loadOne', () => {
      test('should set the status to loading', () => {
        expect(
          reducer(undefined, WeatherViewPageActions.loadOne())
        ).toMatchSnapshot();
      });
    });

    describe('ViewPageActions.loadOneSuccess', () => {
      test('should set the status to loaded and store the id', () => {
        const weather = weatherFixtures({ id: 1 });

        expect(
          reducer(undefined, WeatherViewPageActions.loadOneSuccess(weather))
        ).toMatchSnapshot();
      });

      test('should reset the error', () => {
        const weather = weatherFixtures({ id: 1 });
        const state = {
          ...initialState,
          error: new Error('an error')
        };

        expect(
          reducer(state, WeatherViewPageActions.loadOneSuccess(weather))
        ).toMatchSnapshot();
      });
    });

    describe('ViewPageActions.loadOneError', () => {
      test('should set the error', () => {
        const error = new Error('an error')

        expect(
          reducer(undefined, WeatherViewPageActions.loadOneError(error))
        ).toMatchSnapshot();
      });
    });

  });

  describe('selector:', () => {

    describe('getLoaded', () => {
      test('should return if the status is set to loaded', () =>{
        const result = getLoaded({
          ...initialState,
          status: 'loaded'
        });

        expect(
          result
        ).toMatchSnapshot();
      });
    });

    describe('getLoading', () => {
      test('should return if the status is set to loading', () =>{
        const result = getLoading({
          ...initialState,
          status: 'loading'
        });

        expect(
          result
        ).toMatchSnapshot();
      });
    });

    describe('getError', () => {
      test('should return if the status is set to error', () =>{
        const result = getError({
          ...initialState,
          status: 'error'
        });

        expect(
          result
        ).toMatchSnapshot();
      });
    });

    describe('getId', () => {
      test('should return the id', () => {
        const result = getId({
          ...initialState,
          id: 111
        });

        expect(
          result
        ).toMatchSnapshot();
      });
    });

  });

});
