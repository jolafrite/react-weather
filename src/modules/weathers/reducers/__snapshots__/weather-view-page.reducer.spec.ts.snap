// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer: Weather View Page action: ViewPageActions.loadOne should set the status to loading 1`] = `
Object {
  "error": null,
  "id": null,
  "status": "loading",
}
`;

exports[`reducer: Weather View Page action: ViewPageActions.loadOneError should set the error 1`] = `
Object {
  "error": [Error: an error],
  "id": null,
  "status": "error",
}
`;

exports[`reducer: Weather View Page action: ViewPageActions.loadOneSuccess should reset the error 1`] = `
Object {
  "error": null,
  "id": 1851632,
  "status": "loaded",
}
`;

exports[`reducer: Weather View Page action: ViewPageActions.loadOneSuccess should set the status to loaded and store the id 1`] = `
Object {
  "error": null,
  "id": 1851632,
  "status": "loaded",
}
`;

exports[`reducer: Weather View Page action: ViewPageActions.selectOne should reset set state 1`] = `
Object {
  "error": null,
  "id": 1,
  "status": "pending",
}
`;

exports[`reducer: Weather View Page action: ViewPageActions.selectOne should set the id 1`] = `
Object {
  "error": null,
  "id": 1,
  "status": "pending",
}
`;

exports[`reducer: Weather View Page action: undefined should result the default state 1`] = `
Object {
  "error": null,
  "id": null,
  "status": "pending",
}
`;

exports[`reducer: Weather View Page selector: getError should return if the status is set to error 1`] = `true`;

exports[`reducer: Weather View Page selector: getId should return the id 1`] = `111`;

exports[`reducer: Weather View Page selector: getLoaded should return if the status is set to loaded 1`] = `true`;

exports[`reducer: Weather View Page selector: getLoading should return if the status is set to loading 1`] = `true`;
